packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts

```diff
@@ -0,0 +1,222 @@
+            this.nodes.set(nodePath, new ProxyTarget(this, cursor));
+        }
+        return this.nodes.get(nodePath)! as ProxyTarget;
+    }
+
+    private handleAfterChange(): void {
+        for (const [key, target] of this.nodes) {
+            const path = target.getPath();
+            if (!this.isValidPath(path)) {
+                this.nodes.delete(key);
+                target.free();
+            }
+        }
+        for (const afterHandler of this.afterHandlers) {
+            afterHandler.call(this);
+        }
+    }
+
+    public registerAfterHandler(afterHandler: EditableTreeContextHandler): void {
+        this.afterHandlers.add(afterHandler);
+    }
+
+    public setNodeValue(path: NodePath, value: Value): boolean {
+        return this.runTransaction((editor) => editor.setValue(path, value));
+    }
+
+    public insertNode(path: NodePath, newNodeCursor: ITreeCursor): boolean {
+        return this.runTransaction((editor) => editor.insert(path, newNodeCursor));
+    }
+
+    public deleteNode(path: NodePath, count: number): boolean {
+        return this.runTransaction((editor) => editor.delete(path, count));
+    }
+
+    private runTransaction(f: (editor: SequenceEditBuilder) => void): boolean {
+        assert(this.tree !== undefined, "Transaction-based editing requires `SharedTree` instance");
+        // TODO: currently we can't rely on `invalidateDependents`, since it happens only in forest's `beforeChange`,
+        // and before that delta is applied in `AnchorSet`, so all nodes should already have their anchors allocated.
+        this.prepareForEdit();
+        const result = this.tree.runTransaction((forest: IForestSubscription, editor: SequenceEditBuilder) => {
+            f(editor);
+            return TransactionResult.Apply;
+        });
+        if (result === TransactionResult.Apply) {
+            // TODO: remove as soon as "after change" notification will be implemented in SharedTree
+            this.handleAfterChange();
+            return true;
+        }
+        return false;
+    }
+
+    // this is a workaround to cleanup nodes under deleted parents
+    private isValidPath(path: UpPath | undefined): boolean {
+        let _path = path;
+        try {
+            while (_path?.parent !== undefined) {
+                _path = _path?.parent;
+            }
+        } catch {
+            return false;
+        }
+        return _path !== undefined;
+    }
+}
+
+/**
+ * A simple API for a Forest to interact with the tree.
+ *
+ * @param forest - a Forest to interact with.
+ * @param tree - a SharedTree to handle transactional editing, not required in read-only usecases.
+ * @returns {@link EditableTreeContext} which is used manage the cursors and anchors within the EditableTrees:
+ * This is necessary for supporting using this tree across edits to the forest, and not leaking memory.
+ */
+export function getEditableTreeContext(forest: IEditableForest, tree?: ISharedTree): EditableTreeContext {
```

Comment:
This should not depend on the ISharedTree type. Doing so creates a cyclic dependency and violates our layering. See the architecture section of the package readme.

Reply:
I update it now here, but it has to be discussed.